### ISSUE TITLE
(VB) Conditional access expression in other contexts where expression value is ignored

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#864]
Corresponding to C# issue #74, there are the following expression-lambda-like forms in VB:

```vb
Imports System

Class Program(Of T)
    Sub Main(a As Program(Of T))
        Dim l As Func(Of Task) = Async Function() a?.M() ' should work
        Dim m As Action = Async Sub() a?.M() ' should work
    End Sub
    Function M() As T
        Return Nothing
    End Function
End Class
```

(where the result of M() is an unconstrained type parameter)

The first of these (assignment to `l`) is currently not accepted by the VB compiler, though it should be.

**Comment by gafter** [comment]
This issue was recommended by @AlekseyTs in #799

**Comment by gafter** [comment]
Removing milestone so it can be prioritized in triage.

**Comment by gafter** [comment]
This is not so easy to fix, as it runs against VB language constraints. Specifically, there is a language constraint that the body of a single-line Function lambda must be a value. C# does not have any equivalent constraint, which is why it was straightforward to fix corresponding issue #74.

The equivalent of the following works in C#, but not VB:

```vb
Imports System

Class Program(Of T)
    Sub Main(a As Program(Of T))
        Dim l1 As Func(Of Task) = Async Function() a.VoidThing() ' should work??
    End Sub
    Sub VoidThing()
    End Sub
End Class
```

Unless VB accepts the above, making a special case for the situation requested in this issue is not so easy. And doesn't make much sense, either.

I propose moving this to a feature request for the next major version of the VB language.

@AlekseyTs @ljw1004 what do you think?

**Comment by gafter** [comment]
FYI, this was the text of the unit test I was working on for this:

```vb
Imports System
Imports System.Threading.Tasks

Class Program(Of T)
#Disable Warning BC42356
    Public Shared Sub K(a As Program(Of T), i As Integer)
        Dim l As Func(Of Task) = Async Function() a?.M(i) ' should work
        l().Wait(1) ' should have run synchronously
        Dim m As Action = Async Sub() a?.M(i+1) ' should work
        m()
    End Sub
    Function M(i As Integer) As T
        Console.Write(GetType(T))
        Console.Write(" ")
        Console.WriteLine(i)
        Return Nothing
    End Function
End Class
Module X
    Public Sub Main()
        Program(Of Integer).K(Nothing, 1)
        Program(Of Integer).K(New Program(Of Integer), 3)
        Program(Of String).K(Nothing, 5)
        Program(Of String).K(New Program(Of String), 7)
    End Sub
End Module
```

**Comment by AlekseyTs** [comment]
I feel that the following two statements should be equivalent:
```VB
        Dim l1 As Func(Of Task) = Async Function() a.VoidThing() ' should work??

        Dim l2 As Func(Of Task) = Async Function()
                                      a.VoidThing() ' works
                                  End Function
```

Today they are not. This does look like a feature request to me.

**Comment by AnthonyDGreen** [comment]
We discussed this when we introduced Async. Because in VB = can mean either assignment or equality it's critical to have clear rules with single-line how to interpret the body.

```VB.NET
Function() a = b ' Comparison
Sub() a = b ' Assignment
```
When we introduced Async there were now the potential for three kinds of lambdas:

```VB.NET
Async Function() a = Await b ' Comparison, returns Task(Of Boolean)
Async Function() a = Await b ' Assignment, returns Task
Async Sub() a = Await b ' Assignment, returns Void
```

This was one of the last issues we discussed when debating whether the representation in VB of a task-returning async method should be Async Sub or Async Function As Task. We made the decision then to preserve the existing rule that the body of a single-line function lambda *must* be an expression and ```Async Function() a = Await b``` must always parse as an expression, making the middle case (Async assignment) inexpressible in a single-line. Put another way - there are no Task-returning single-line function lambdas in Visual Basic, by design.
